src/core/server/http/http_server.ts

```diff
@@ -163,43 +163,54 @@ export class HttpServer {
         const validate = isSafeMethod(route.method) ? undefined : { payload: true };
         const { authRequired, tags, body = {}, timeout } = route.options;
         const { accepts: allow, maxBytes, output, parse } = body;
-        // Hapi does not allow timeouts on payloads to be specified for 'head' or 'get' requests
-        const payloadTimeout = isSafeMethod(route.method) || timeout == null ? undefined : timeout;
 
         const kibanaRouteState: KibanaRouteState = {
           xsrfRequired: route.options.xsrfRequired ?? !isSafeMethod(route.method),
         };
 
+        // Works around https://github.com/hapijs/hapi/issues/4122 until v20
+        const fakeSocketTimeout = timeout?.payload ? timeout.payload + 1 : undefined;
+
         this.server.route({
           handler: route.handler,
           method: route.method,
           path: route.path,
           options: {
             auth: this.getAuthOption(authRequired),
             app: kibanaRouteState,
+            ext: {
+              onPreAuth: {
+                method: (request, reply) => {
+                  if (timeout?.idleSocket) {
+                    request.raw.req.socket.setTimeout(timeout?.idleSocket);
+                  } else if (fakeSocketTimeout) {
+                    request.raw.req.socket.setTimeout(this.config!.socketTimeout || 0);
+                  }
+
+                  return reply.continue;
+                },
+              },
+            },
             tags: tags ? Array.from(tags) : undefined,
             // TODO: This 'validate' section can be removed once the legacy platform is completely removed.
             // We are telling Hapi that NP routes can accept any payload, so that it can bypass the default
             // validation applied in ./http_tools#getServerOptions
             // (All NP routes are already required to specify their own validation in order to access the payload)
             validate,
-            payload: [allow, maxBytes, output, parse, payloadTimeout].some(
+            payload: [allow, maxBytes, output, parse, timeout?.payload].some(
               (v) => typeof v !== 'undefined'
             )
```

Comment:
(slightly out of scope) Do we really need this `.some` check to return `undefined` instead of an object of undefined values? I would expect Hapi is using Joi under the hood to populate the defaults here anyways.

Reply:
It is... per https://github.com/elastic/kibana/pull/50783#discussion_r347274206 HapiJS doesn't let us set the `payload` option for a `GET` or `HEAD` route definition...